wrappers/glmemshadow.cpp

```diff
@@ -0,0 +1,349 @@
+    mappedSize = size;
+
+    mappedStartPage = start / sPageSize;
+    mappedEndPage = divRoundUp(start + size, sPageSize);
+
+    uint8_t *protectStart = shadowMemory + mappedStartPage * sPageSize;
+    const size_t protectSize = (mappedEndPage - mappedStartPage) * sPageSize;
+
+    // The buffer may have been updated before the mapping.
+    // TODO: handle write only buffers
+    if (flags & GL_MAP_READ_BIT) {
+        memProtect(protectStart, protectSize, MemProtection::READ_WRITE);
+        memcpy(shadowMemory + start, glMemory + start, size);
+    }
+
+    memProtect(protectStart, protectSize, MemProtection::READ_ONLY);
+
+    return shadowMemory + start;
+}
+
+void GLMemoryShadow::unmap(Callback callback)
+{
+    if (isDirty) {
+        commitWrites(callback);
+    }
+
+    {
+        os::unique_lock<os::mutex> lock(mutex);
+
+        auto it = std::find(ctx->dirtyShadows.begin(), ctx->dirtyShadows.end(), this);
+        if (it != ctx->dirtyShadows.end()) {
+            ctx->dirtyShadows.erase(it);
+        }
+    }
+
+    memProtect(shadowMemory, nPages * sPageSize, MemProtection::NO_ACCESS);
+
+    ctx = nullptr;
+    glMemory = nullptr;
+    mappedStart = 0;
+    mappedSize = 0;
+    pagesToDirtyOnConsecutiveWrites = 1;
+}
+
+void GLMemoryShadow::onAddressWrite(uintptr_t addr, size_t page)
+{
+    const size_t relativePage = (addr - reinterpret_cast<uintptr_t>(shadowMemory)) / sPageSize;
+    if (relativePage > 0 && isPageDirty(relativePage - 1)) {
+        pagesToDirtyOnConsecutiveWrites *= 2;
+    } else {
+        pagesToDirtyOnConsecutiveWrites = 1;
+    }
+
+    const size_t endPageToDirty = std::min(relativePage + pagesToDirtyOnConsecutiveWrites, nPages);
+    for (size_t pageToDirty = relativePage; pageToDirty < endPageToDirty; pageToDirty++) {
+        setPageDirty(pageToDirty);
+    }
+
+    memProtect(reinterpret_cast<void*>(page * sPageSize),
+               (endPageToDirty - relativePage) * sPageSize, MemProtection::READ_WRITE);
+}
+
+void GLMemoryShadow::setPageDirty(size_t relativePage)
+{
+    assert(relativePage < nPages);
+    dirtyPages[relativePage / 32] |= 1U << (relativePage % 32);
+
+    if (!isDirty) {
+        ctx->dirtyShadows.push_back(this);
+        isDirty = true;
+    }
+}
+
+bool GLMemoryShadow::isPageDirty(size_t relativePage)
+{
+    assert(relativePage < nPages);
+    return dirtyPages[relativePage / 32] & (1U << (relativePage % 32));
+}
+
+void GLMemoryShadow::commitWrites(Callback callback)
+{
+    os::unique_lock<os::mutex> lock(mutex);
+
+    assert(isDirty);
+
+    uint8_t *shadowSlice = shadowMemory + mappedStartPage * sPageSize;
+    const size_t glStartOffset = mappedStart % sPageSize;
+
+    if (isPageDirty(mappedStartPage)) {
+        const size_t shadowOffset = mappedStart % sPageSize;
+        const size_t size = std::min(sPageSize - glStartOffset, mappedSize);
+
+        memcpy(glMemory, shadowSlice + shadowOffset, size);
+        callback(shadowSlice + shadowOffset, size);
+    }
```

Comment:
I have a feeling that here should be:
`shadowSlice += shadowOffset`

Reply:
For the first dirty page we may need to start copying from `shadowSlice` with an offset depending on the mapping start offset and will copy `pageSize - offset` bytes.

For the next page the copying will start from the start of the page in shadowSlice which is `shadowSlice + pageSize` and would copy into `glOffset + pageSize - glStartOffset`.

So no, there is no need to change `shadowSlice` pointer.

Reply:
yes, you are right